### ISSUE TITLE
Unable to set retry policy for filter.http.ext_authz.v2.ExtAuthz

### ISSUE THREAD

**Issue by TuvaevAndrey** [envoyproxy/envoy#9396]
Hi everyone. In accordance with documentation the only way to configure retry policy for ExtAuthz filter is to use request headers. Here is my config:
```
admin:
  access_log_path: /dev/stdout
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 9901
dynamic_resources:
  ads_config:
    api_type: GRPC
    grpc_services:
      envoy_grpc:
        cluster_name: xds_cluster
  cds_config:
    ads: {}
static_resources:
  listeners:
    - name: main
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8080
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
                stat_prefix: customer
                codec_type: AUTO
                rds:
                  route_config_name: local_route
                  config_source:
                    ads: {}
                http_filters:
                  - name: envoy.cors
                  - name: envoy.filters.http.jwt_authn
                    config:
                      providers:
                        local_provider:
                          forward: true
                          local_jwks:
                            inline_string: |
                              {
                                "keys": [
                                  {
                                    "kty":"RSA",
                                    "e":"AQAB",
                                    "kid":"38b89b33-9680-4869-a142-56d91ef0f0e3",
                                    "n":"some_value"
                                  }
                                ],
                              }
                      rules:
                      - match:
                          prefix: "/"
                        requires: { provider_name: local_provider }
                  - name: envoy.ext_authz
                    typed_config:
                      "@type": type.googleapis.com/envoy.config.filter.http.ext_authz.v2.ExtAuthz
                      grpc_service:
                        timeout: 3s
                        envoy_grpc:
                          cluster_name: acl_cluster
                        initial_metadata:
                          - key: x-envoy-retry-grpc-on
                            value: "unavailable"
                          - key: x-envoy-attempt-count
                            value: "3"
                          - key: x-envoy-max-retries
                            value: "3"
                  - name: envoy.filters.http.header_to_metadata
                    config:
                      request_rules:
                        - header: instance-id
                          on_header_present:
                            metadata_namespace: envoy.lb
                            key: instanceId
                            type: STRING
                          remove: true
                  - name: envoy.router
  clusters:
    - name: xds_cluster
      connect_timeout: 3s
      type: STRICT_DNS
      dns_lookup_family: V4_ONLY
      http2_protocol_options: {}
      lb_policy: ROUND_ROBIN
      load_assignment:
        cluster_name: xds_cluster
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: docker.for.mac.localhost
                      port_value: 8888
    - name: acl_cluster
      connect_timeout: 3s
      type: STRICT_DNS
      dns_lookup_family: V4_ONLY
      http2_protocol_options: {}
      lb_policy: ROUND_ROBIN
      health_checks:
        timeout: 2s
        interval: 5s
        no_traffic_interval: 60s
        unhealthy_threshold: 2
        healthy_threshold: 3
        event_log_path: /dev/stdout
        grpc_health_check: {}
      load_assignment:
        cluster_name: acl_cluster
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: docker.for.mac.localhost
                      port_value: 8889

```
As you may see I put 'x-envoy-retry-grpc-on' header inside 'initial_metadata' in 'grpc_service' section ExtAuthz filter's configuration but envoy still doesn't retries requests to disabled auth service which returns "14" grpc code.

```
[2019-12-18 09:54:09.467][15][debug][filter] [source/extensions/filters/http/jwt_authn/authenticator.cc:102] Jwt authentication starts
[2019-12-18 09:54:09.468][15][debug][filter] [source/extensions/filters/http/jwt_authn/authenticator.cc:243] Jwt authentication completed with: OK
[2019-12-18 09:54:09.468][15][debug][filter] [source/extensions/filters/http/jwt_authn/filter.cc:58] Called Filter : check complete 0
[2019-12-18 09:54:09.468][15][trace][http] [source/common/http/conn_manager_impl.cc:857] [C4][S8420969345045605427] decode headers called: filter=0x2c3be50 status=0
[2019-12-18 09:54:09.468][15][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:72] [C4][S8420969345045605427] ext_authz filter calling authorization server
[2019-12-18 09:54:09.468][15][debug][router] [source/common/router/router.cc:401] [C0][S18049916063950310141] cluster 'acl_cluster' match for URL '/envoy.service.auth.v2.Authorization/Check'
[2019-12-18 09:54:09.468][15][debug][router] [source/common/router/router.cc:514] [C0][S18049916063950310141] router decoding headers:
':method', 'POST'
':path', '/envoy.service.auth.v2.Authorization/Check'
':authority', 'acl_cluster'
':scheme', 'http'
'te', 'trailers'
'grpc-timeout', '3000m'
'content-type', 'application/grpc'
'x-envoy-attempt-count', '3'
'x-envoy-internal', 'true'
'x-forwarded-for', '172.18.0.2'
'x-envoy-expected-rq-timeout-ms', '3000'

[2019-12-18 09:54:09.468][15][debug][client] [source/common/http/codec_client.cc:26] [C53] connecting
[2019-12-18 09:54:09.468][15][debug][connection] [source/common/network/connection_impl.cc:702] [C53] connecting to 192.168.65.2:8889
[2019-12-18 09:54:09.468][15][debug][connection] [source/common/network/connection_impl.cc:711] [C53] connection in progress
[2019-12-18 09:54:09.468][15][debug][http2] [source/common/http/http2/codec_impl.cc:726] [C53] setting stream-level initial window size to 268435456
[2019-12-18 09:54:09.468][15][debug][http2] [source/common/http/http2/codec_impl.cc:748] [C53] updating connection-level initial window size to 268435456
[2019-12-18 09:54:09.468][15][debug][pool] [source/common/http/conn_pool_base.cc:20] queueing request due to no available connections
[2019-12-18 09:54:09.468][15][trace][router] [source/common/router/router.cc:1393] [C0][S18049916063950310141] buffering 1820 bytes
[2019-12-18 09:54:09.468][15][trace][http] [source/common/http/conn_manager_impl.cc:857] [C4][S8420969345045605427] decode headers called: filter=0x2a1ebe0 status=4
[2019-12-18 09:54:09.468][15][trace][http] [source/common/http/http1/codec_impl.cc:387] [C4] parsed 1578 bytes
[2019-12-18 09:54:09.468][15][trace][connection] [source/common/network/connection_impl.cc:288] [C4] readDisable: enabled=true disable=true
[2019-12-18 09:54:09.468][15][trace][connection] [source/common/network/connection_impl.cc:456] [C4] socket event: 2
[2019-12-18 09:54:09.468][15][trace][connection] [source/common/network/connection_impl.cc:541] [C4] write ready
[2019-12-18 09:54:09.469][15][trace][connection] [source/common/network/connection_impl.cc:456] [C53] socket event: 3
[2019-12-18 09:54:09.469][15][trace][connection] [source/common/network/connection_impl.cc:541] [C53] write ready
[2019-12-18 09:54:09.469][15][debug][connection] [source/common/network/connection_impl.cc:559] [C53] delayed connection error: 111
[2019-12-18 09:54:09.469][15][debug][connection] [source/common/network/connection_impl.cc:188] [C53] closing socket: 0
[2019-12-18 09:54:09.469][15][debug][client] [source/common/http/codec_client.cc:82] [C53] disconnect. resetting 0 pending requests
[2019-12-18 09:54:09.469][15][debug][pool] [source/common/http/http2/conn_pool.cc:149] [C53] client disconnected
[2019-12-18 09:54:09.469][15][debug][router] [source/common/router/router.cc:868] [C0][S18049916063950310141] upstream reset: reset reason connection failure
[2019-12-18 09:54:09.469][15][debug][http] [source/common/http/async_client_impl.cc:91] async http request response headers (end_stream=true):
':status', '200'
'content-type', 'application/grpc'
'grpc-status', '14'
'grpc-message', 'upstream connect error or disconnect/reset before headers. reset reason: connection failure'

[2019-12-18 09:54:09.469][15][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:218] [C4][S8420969345045605427] ext_authz filter rejected the request with an error. Response status code: 403
[2019-12-18 09:54:09.469][15][debug][http] [source/common/http/conn_manager_impl.cc:1167] [C4][S8420969345045605427] Sending local reply with details ext_authz_error
[2019-12-18 09:54:09.469][15][trace][http] [source/common/http/conn_manager_impl.cc:1254] [C4][S8420969345045605427] encode headers called: filter=0x29702c0 status=0
[2019-12-18 09:54:09.469][15][trace][http] [source/common/http/conn_manager_impl.cc:1254] [C4][S8420969345045605427] encode headers called: filter=0x2970e80 status=0
[2019-12-18 09:54:09.469][15][trace][http] [source/common/http/conn_manager_impl.cc:1254] [C4][S8420969345045605427] encode headers called: filter=0x2971040 status=0
[2019-12-18 09:54:09.469][15][trace][http] [source/common/http/conn_manager_impl.cc:1254] [C4][S8420969345045605427] encode headers called: filter=0x29709c0 status=0
[2019-12-18 09:54:09.469][15][debug][http] [source/common/http/conn_manager_impl.cc:1359] [C4][S8420969345045605427] encoding headers via codec (end_stream=true):
':status', '403'
'date', 'Wed, 18 Dec 2019 09:54:09 GMT'
'server', 'envoy'

[2019-12-18 09:54:09.469][15][trace][connection] [source/common/network/connection_impl.cc:392] [C4] writing 97 bytes, end_stream false
[2019-12-18 09:54:09.469][15][debug][filter] [source/extensions/filters/http/jwt_authn/filter.cc:24] Called Filter : onDestroy
[2019-12-18 09:54:09.469][15][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=1)
[2019-12-18 09:54:09.469][15][trace][connection] [source/common/network/connection_impl.cc:288] [C4] readDisable: enabled=false disable=false
[2019-12-18 09:54:09.469][15][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=2)
[2019-12-18 09:54:09.469][15][trace][main] [source/common/event/dispatcher_impl.cc:158] item added to deferred deletion list (size=3)
[2019-12-18 09:54:09.469][15][debug][pool] [source/common/http/http2/conn_pool.cc:171] [C53] destroying primary client
```

Looking forward to hearing from you, let me know if I have missed some parts in the documentation.

**Comment by TuvaevAndrey** [comment]
Guys?

**Comment by dio** [comment]
@TuvaevAndrey, sorry for the delayed response.

Could you help to try the following?

```yaml
          - name: envoy.ext_authz
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.ext_authz.v2.ExtAuthz
              grpc_service:
                envoy_grpc:
                  cluster_name: service_ext_authz
                initial_metadata:
                - key: x-envoy-retry-on
                  value: "5xx"

```

I ran this:

```
$ envoy -c config.yaml --component-log-level router:debug
... snip ...
[2020-01-19 00:03:13.373][9767][debug][router] [source/common/router/router.cc:474] [C0][S2841941481599445446] cluster 'service_ext_authz' match for URL '/envoy.service.auth.v2.Authorization/Check'
[2020-01-19 00:03:13.373][9767][debug][router] [source/common/router/router.cc:614] [C0][S2841941481599445446] router decoding headers:
':method', 'POST'
':path', '/envoy.service.auth.v2.Authorization/Check'
':authority', 'service_ext_authz'
':scheme', 'http'
'te', 'trailers'
'grpc-timeout', '200m'
'content-type', 'application/grpc'
'x-envoy-internal', 'true'
'x-forwarded-for', '10.148.0.3'
'x-envoy-expected-rq-timeout-ms', '200'

[2020-01-19 00:03:13.374][9767][debug][router] [source/common/router/router.cc:1711] [C0][S2841941481599445446] pool ready
[2020-01-19 00:03:13.375][9767][debug][router] [source/common/router/router.cc:990] [C0][S2841941481599445446] upstream reset: reset reason connection termination
[2020-01-19 00:03:13.375][9767][debug][router] [source/common/router/router.cc:1357] [C0][S2841941481599445446] performing retry
[2020-01-19 00:03:13.392][9767][debug][router] [source/common/router/router.cc:1711] [C0][S2841941481599445446] pool ready
[2020-01-19 00:03:13.393][9767][debug][router] [source/common/router/router.cc:990] [C0][S2841941481599445446] upstream reset: reset reason connection termination
```

The content of `config.yaml`:

```yaml
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 127.0.0.1
      port_value: 9901
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: service_google
          http_filters:
          - name: envoy.ext_authz
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.ext_authz.v2.ExtAuthz
              grpc_service:
                envoy_grpc:
                  cluster_name: service_ext_authz
                initial_metadata:
                - key: x-envoy-retry-on
                  value: "5xx"
          - name: envoy.router
  clusters:
  - name: service_ext_authz
    connect_timeout: 5s
    type: STATIC
    http2_protocol_options: {}
    load_assignment:
      cluster_name: service_ext_authz
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 127.0.0.2 # Always gives 503
                port_value: 3000

  - name: service_google
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_google
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.google.com
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
        sni: www.google.com
```

**Comment by klarose** [comment]
I'm planning on trying this out in my own service today. I wouldn't mind confirming my understanding of the solution.

@dio I'm guessing the difference in your config compared to that in the main issue post is that you are retrying on http 5xx, rather than grpc error codes. Presumably this fixes the problem because a failure to connect to the upstream would give a 5xx, and not a grpc error code, since the grpc connection was never actually established.

**Comment by dio** [comment]
@klarose yes, we can try to test on when gRPC connection happens too. If that doesn't trigger retry then it is a bug.

**Comment by klarose** [comment]
The approach of setting the retry headers for me. However, I had to use an envoy_grpc endpoint, not a google_grpc endpoint. 

This actually caused some issues for me, since I'm using istio: the cluster created by istio has `|` in its name. envoy_grpc uses the cluster name as the authority in the http2 portion of the grpc request. The upstream envoy rejects that as invalid. I plan on submitting an issue for that, though if you have any suggestions I'm all ears. :)


FWIW, it was important to set a few other things for this to be fully resilient. I needed to make sure the upstream cluster had a connection timeout less than the ext_authz timeout. Otherwise, an endpoint which disappeared (e.g. the node hosting it crashed/rebooted) would cause the request to time out without a retry if that request triggered a new connection attempt.

I also had to set the `x-envoy-upstream-rq-per-try-timeout-ms` header so that if an established connection experienced the connectivity issues, we wouldn't time out waiting for a reply. Note: you want to set the per-try timeout. The other timeout header has the same semantics as the ext_authz timeout.
